feat: ValidatorKeystore implementation for high-availability signer

spypsy opened the pull request:
Fixes [A-354](https://linear.app/aztec-labs/issue/A-354/integrate-slash-protection-in-signing-path)

Integrates the existing validator HA signer into validator-client via a new `HAKeyStore` wrapper, adds type-safe duty identifiers using discriminated unions, and includes integration testing with multiple validator nodes.

## What's New

### 1. HAKeyStore Implementation

**New File**: `yarn-project/validator-client/src/key_store/ha_key_store.ts`

Transparent wrapper around `ExtendedValidatorKeyStore` that adds HA coordination:

- Drop-in replacement for any `ExtendedValidatorKeyStore` implementation
- AUTH_REQUEST duties bypass HA coordination (safe to sign multiple times)
- Signs with all addresses concurrently using `Promise.allSettled()`, filtering out already-signed duties

```typescript
// With HA protection
const sigs = await haKeyStore.signTypedData(typedData, {
  slot: 100n,
  blockNumber: 50n,
  blockIndexWithinCheckpoint: 0,
  dutyType: DutyType.BLOCK_PROPOSAL,
});

// Without HA protection (AUTH_REQUEST)
const sigs = await haKeyStore.signTypedData(typedData, {
  dutyType: DutyType.AUTH_REQUEST,
});
```

### 2. Type-Safe Duty Identifiers

**Modified**: `yarn-project/validator-ha-signer/src/db/types.ts`

Replaced optional `blockIndexWithinCheckpoint?: number` with discriminated unions:

```typescript
type DutyIdentifier =
  | BlockProposalDutyIdentifier // MUST have blockIndexWithinCheckpoint >= 0
  | OtherDutyIdentifier; // Does NOT have this field

interface BlockProposalDutyIdentifier {
  validatorAddress: EthAddress;
  slot: SlotNumber;
  blockIndexWithinCheckpoint: number; // Required
  dutyType: DutyType.BLOCK_PROPOSAL;
}
```

Enforces correctness at compile-time and prevents invalid database primary keys.

### 3. Extended Duty Type Support

Added: `CHECKPOINT_PROPOSAL`, `GOVERNANCE_VOTE`, `SLASHING_VOTE`, `AUTH_REQUEST`

### 4. Integration Test

**New File**: `yarn-project/validator-client/src/validator.ha.integration.test.ts`

E2E test with 5 validators sharing a PostgreSQL database. Tests concurrent block proposals, attestations, checkpoint proposals, and slashing protection.

## Key Changes

### Cleanup Timeout Configuration

Made `maxStuckDutiesAgeMs` dynamically computed from actual slot duration:

```typescript
const haConfig = {
  ...config,
  maxStuckDutiesAgeMs:
    config.maxStuckDutiesAgeMs ??
    epochCache.getL1Constants().slotDuration * 2 * 1000,
};
```

### SigningContext Now Required

All `ValidatorKeyStore` signing methods now require `SigningContext` parameter. Updated `LocalKeyStore`, `NodeKeystoreAdapter`, `Web3SignerKeyStore`, and new `HAKeyStore`.

### Database Schema

Added `block_index_within_checkpoint` to primary key to support multiple block proposals per slot:

```sql
PRIMARY KEY (validator_address, slot, duty_type, block_index_within_checkpoint)
```

## Configuration

```bash
VALIDATOR_HA_DATABASE_URL=postgresql://user:pass@host:port/db
VALIDATOR_HA_SIGNING_ENABLED=true
VALIDATOR_HA_NODE_ID=validator-node-1
VALIDATOR_HA_POLLING_INTERVAL_MS=100        # Optional
VALIDATOR_HA_SIGNING_TIMEOUT_MS=3000        # Optional
VALIDATOR_HA_MAX_STUCK_DUTIES_AGE_MS=144000 # Optional (defaults to 2× slot duration)
```

## Notes

- HA is opt-in via `VALIDATOR_HA_SIGNING_ENABLED=true`
- Non-HA validators work unchanged
- Lock-free read operations
- Atomic lock acquisition using `INSERT ... ON CONFLICT` with retry logic (10ms, 20ms, 30ms backoff) to handle PostgreSQL transaction visibility edge cases in high-concurrency scenarios

PhilWindle commented on yarn-project/validator-client/src/duties/validation_service.ts:
```diff
@@ -73,25 +87,73 @@ export class ValidationService {
+            }
+          : undefined;
+
+        const sig = await this.keyStore.signMessageWithAddress(attestor, buf, context);
```
This could be a performance concern. If someone had a lot of validators in the committee.

alexghr replied:
I'd expect this to just be the set of attesters part of the current committee, at most 24. If it takes 300ms to produce a HA signature then producing 24 of them would take 7.2s (but no single machine would have all 24 committee members)

Edit: 300ms is just a random number. I didn't benchmark how long signing through a HA setup could take

alexghr replied:
Another question on this: does this above sign request block if the duty is in `SIGNING` mode?

PhilWindle commented on yarn-project/validator-client/src/key_store/ha_key_store.ts:
```diff
@@ -0,0 +1,226 @@
+
+  constructor(
+    private readonly baseKeyStore: ExtendedValidatorKeyStore,
+    private readonly haSigner: ValidatorHASigner | null,
```
Why are we using `null` in this class?

spypsy replied:
no longer handles **not** having a `haSigner`

alexghr commented on yarn-project/sequencer-client/src/sequencer/checkpoint_proposal_job.ts:
```diff
@@ -207,21 +208,67 @@ export class CheckpointProposalJob {
+          blockProposalOptions,
+        );
+      } catch (err) {
+        if (err instanceof DutyAlreadySignedError) {
```
Could these two errors be handled in a higher function to avoid error-handling duplication at this point (proposal creation) and later, at signing?

spypsy replied:
updated to handle inside `proposeCheckpoint`, wrapping `buildBlocksForCheckpoint` which is the first operation that requires a signature

alexghr commented on yarn-project/validator-client/src/duties/validation_service.ts:
```diff
@@ -73,25 +87,73 @@ export class ValidationService {
+          this.log.debug(
+            `Attestation for slot ${proposal.slotNumber} by ${attestor} already signed by another High-Availability node`,
+          );
+          // Continue with remaining attestors
+        } else {
+          throw error;
+        }
```
Shouldn't it continue with the other attestors all the time?

spypsy replied:
given this replaces a `Promise.all`, we were already throwing at any error within the promises. 
But I may go back to using `Promise.all`, I'm looking into this now

alexghr commented on yarn-project/validator-client/src/duties/validation_service.ts:
```diff
@@ -73,25 +87,73 @@ export class ValidationService {
+    slot: SlotNumber,
+    blockNumber: BlockNumber | CheckpointNumber,
   ): Promise<Signature> {
+    const context: SigningContext = {
+      slot,
+      blockNumber,
+      dutyType: DutyType.ATTESTATIONS_AND_SIGNERS,
+    };
```
I'm wondering if we could just simplify this context down to only `dutyType` and `slot`? It's probably not possible now that there's >1 block per slot.

spypsy replied:
yeah this was added for future-proofing

PhilWindle commented on yarn-project/sequencer-client/src/sequencer/checkpoint_proposal_job.ts:
```diff
@@ -174,12 +175,45 @@ export class CheckpointProposalJob {
+          return undefined;
+        }
+        if (err instanceof SlashingProtectionError) {
+          this.log.warn(
```
If we are defining this as normal, it shouldn't be a warning should it? This will scare people.

PhilWindle commented on yarn-project/sequencer-client/src/sequencer/checkpoint_proposal_job.ts:
```diff
@@ -226,7 +261,32 @@ export class CheckpointProposalJob {
+          return undefined;
+        }
+        if (err instanceof SlashingProtectionError) {
+          this.log.warn(`Attestations signature for slot ${this.slot} blocked by slashing protection`, {
```
Same here

PhilWindle commented on yarn-project/validator-client/src/duties/validation_service.ts:
```diff
@@ -73,25 +143,75 @@ export class ValidationService {
+          ? {
+              slot: proposal.slotNumber,
+              blockNumber,
+              blockIndexWithinCheckpoint: -1, // -1 indicates not applicable (attestation, not a block proposal)
```
I'm wondering if this should be optional rather than set to -1 whenevre it's not a block proposal being signed.

spypsy replied:
yeah would actually make sense to make it optional in `SigningContext` and just use -1 at the bottom DB layer

spypsy replied:
updated to have strict type-checking for different duty types

PhilWindle commented on yarn-project/validator-client/src/validator.ts:
```diff
@@ -457,7 +475,19 @@ export class ValidatorClient extends (EventEmitter as new () => WatcherEmitter)
     this.log.info(`Assembling checkpoint proposal for slot ${header.slotNumber}`);
-    return this.createBlockProposal(0 as BlockNumber, header, archive, txs, proposerAddress, options);
+    // Derive checkpoint number from the header's slot
+    const checkpointNumber = header.slotNumber as unknown as CheckpointNumber;
```
This isn't right. Slot number !== Checkpoint number.

spypsy replied:
nope, missed this one. fixing

PhilWindle commented on yarn-project/validator-ha-signer/src/errors.ts:
```diff
@@ -10,8 +12,9 @@ import type { DutyType } from './db/types.js';
-    public readonly slot: bigint,
+    public readonly slot: SlotNumber,
     public readonly dutyType: DutyType,
+    public readonly blockIndexWithinCheckpoint: number,
```
I'm wondering if we should make a new branded type.

spypsy replied:
might need @spalladino input on this, I know he has more work on checkpoints in progress

spalladino replied:
I have a TODO for that, heh

alexghr commented on yarn-project/validator-client/src/duties/validation_service.ts:
```diff
@@ -73,25 +142,79 @@ export class ValidationService {
+      } else {
+        const error = result.reason;
+        if (error instanceof DutyAlreadySignedError || error instanceof SlashingProtectionError) {
+          this.log.debug(
```
I'd promote this to INFO/WARN level. Operators running in HA mode will want to know that their setup is working correctly and that when a node produces a signature the other ones notice it and don't try to sign.

alexghr commented on yarn-project/validator-client/src/key_store/interface.ts:
```diff
@@ -26,17 +29,45 @@ export interface ValidatorKeyStore {
+   * @param context - Optional signing context for HA slashing protection
+   * @returns signatures (when context provided with HA, only successfully claimed signatures are returned)
+   */
+  signTypedData(typedData: TypedDataDefinition, context?: SigningContext): Promise<Signature[]>;
```
Could we make `context` mandatory?

spypsy replied:
started doing this but there are cases where we don't want to go through high-availability protection, like slashing & governance votes. We can either pass context for those as well and handle them specifically depending on `DutyType` or keep as is where we don't pass context for signatures that don't need HA protection

alexghr commented on yarn-project/validator-client/src/key_store/ha_key_store.ts:
```diff
@@ -0,0 +1,275 @@
+    // No context provided - sign directly
+    if (!context) {
+      return this.baseKeyStore.signTypedData(typedData);
+    }
```
I think this is dangerous. I'd much rather this signer crash in this case. Without a context it can't do its job so rather than doing something unsafe it should throw an error.

spypsy replied:
refactoring to make `context` required

spypsy replied:
update on [this comment](https://github.com/AztecProtocol/aztec-packages/pull/19094#discussion_r2685641194)

alexghr commented on yarn-project/validator-client/src/key_store/ha_key_store.ts:
```diff
@@ -0,0 +1,275 @@
+          return false;
+        }
+        // Re-throw unexpected errors
+        throw result.reason;
```
Throwing from inside `filter` feels weird. I wonder what happens to the stack trace. It might be better to `throw new Error('Unexpected error', { cause: result.reason });` to also make sure we're throwing errors

Later edit: the stack trace is maintained and points to the source of the error rather than the `filter` callback

alexghr commented on yarn-project/validator-client/src/key_store/ha_key_store.ts:
```diff
@@ -0,0 +1,275 @@
+   * Logs expected HA errors (already signed) at appropriate levels.
+   * Re-throws unexpected errors.
+   */
+  private logSigningError(error: unknown, context: SigningContext) {
```
This function re-throws the error. I suggest we rename it because `log` sounds too safe.

```suggestion
  private processSigningError(error: unknown, context: SigningContext) {
```

alexghr commented on yarn-project/validator-client/src/key_store/interface.ts:
```diff
@@ -79,4 +110,10 @@ export interface ExtendedValidatorKeyStore extends ValidatorKeyStore {
+   * Is high-availability key store
+   * @returns true if the key store is a high-availability key store
+   */
+  isHAKeyStore(): this is HAKeyStore;
```
Having an `isXYZImplementation()` directly on an interface feels like the wrong approach.

spypsy replied:
open to suggestions 😬
I was initially checking if start/stop methods are available but that feels even worse.
Guess we can just make start/stop part of the interface and have it be a no-op for the non-HA keystores

alexghr commented on yarn-project/validator-client/src/validator.test.ts:
```diff
@@ -701,4 +716,15 @@ describe('ValidatorClient', () => {
+
+  describe('lifecycle methods', () => {
+    it('should run start() / stop() on the HA key store', async () => {
+      (validatorClient as any).config.haSigningEnabled = true;
+      (validatorClient as any).keyStore = haKeyStore;
+      await validatorClient.start();
+      expect(haKeyStore.start).toHaveBeenCalledTimes(1);
+      await validatorClient.stop();
+      expect(haKeyStore.stop).toHaveBeenCalledTimes(1);
```
Is is the validator's responsibility to know it's connected to a HA signer and start it or should the HA signer be an orthogonal service that gets started separately and a handle be given to the validator? I'm in favour of the later.

alexghr commented on yarn-project/validator-client/src/validator.ts:
```diff
@@ -170,8 +173,14 @@ export class ValidatorClient extends (EventEmitter as new () => WatcherEmitter)
+    let validatorKeyStore: ExtendedValidatorKeyStore = NodeKeystoreAdapter.fromKeyStoreManager(keyStoreManager);
+    if (config.haSigningEnabled) {
+      const { signer } = await createHASigner(config);
+      validatorKeyStore = new HAKeyStore(validatorKeyStore, signer);
```
I think this is where the HA key store should be started, in the factory method for the validator.

spypsy replied:
Could do but it does stray away a bit from our pattern of starting all underlying services during the consumer's `start` / `stop` methods

alexghr commented on yarn-project/validator-ha-signer/src/db/postgres.test.ts:
```diff
@@ -507,26 +543,135 @@ describe('PostgresSlashingProtectionDatabase', () => {
+      let callCount = 0;
+
+      // Mock pool.query to return no rows on first call, then rows on second call
+      const originalQuery = pool.query.bind(pool);
+      jest.spyOn(pool, 'query').mockImplementation(async (...args: any[]) => {
+        callCount++;
+        if (callCount === 1) {
+          // First call: simulate race condition - no rows returned
+          return { rows: [] };
+        }
+        // Subsequent calls: return actual result
+        return await originalQuery(...args);
+      });
```
If I understand the function correctly a better test (compared to mocking) would t be to insert the row into the database, call `tryInsertOrGetExisting` and while it is running, delete the row to prove that it locks when it is possible

spypsy replied:
that's not really the race condition here. The issue is when the the query gets a conflict on `INSERT` but `SELECT` returns nothing (bc a concurrent process is still inserting). 
I couldn't really reproduce with `pglite`, it only occured when connecting to a 'real' postgres instance, hence the mocking

alexghr commented on yarn-project/validator-ha-signer/src/db/postgres.test.ts:
```diff
@@ -507,26 +543,135 @@ describe('PostgresSlashingProtectionDatabase', () => {
+      jest.restoreAllMocks();
+    });
+
+    it('should retry multiple times if needed', async () => {
```
Doesn't this test run the exact same check but.. twice? In that case the previous test should be parameterised

alexghr commented on yarn-project/validator-ha-signer/src/db/postgres.test.ts:
```diff
@@ -507,26 +543,135 @@ describe('PostgresSlashingProtectionDatabase', () => {
+      jest.restoreAllMocks();
+    });
+
+    it('should throw error after all retries are exhausted', async () => {
```
How many retries?

alexghr commented on yarn-project/validator-ha-signer/src/db/postgres.test.ts:
```diff
@@ -571,4 +718,325 @@ describe('PostgresSlashingProtectionDatabase', () => {
+      const spDb = new PostgresSlashingProtectionDatabase(pool);
+
+      // Insert a duty first
+      const insertResult = await spDb.tryInsertOrGetExisting({
```
Does the database get cleared between tests?

spypsy replied:
yep 
```
  beforeEach(async () => {
    db = new PGlite();

    await setupTestSchema(db);
  });

  afterEach(async () => {
    await db.close();
  });
```

alexghr commented on yarn-project/validator-ha-signer/src/db/postgres.ts:
```diff
@@ -72,24 +74,46 @@ export class PostgresSlashingProtectionDatabase implements SlashingProtectionDat
+          lockToken,
+        ]);
+
+        // Throw error if no rows to trigger retry
+        if (queryResult.rows.length === 0) {
+          throw new Error('INSERT_OR_GET_DUTY returned no rows');
+        }
```
The query _must_ return >=1 row, that's how the SQL is written. Did this check get added to check retry logic in the unit tests? If so I think we should look for a different way to trigger retries (maybe mock network errors?)

alexghr commented on yarn-project/validator-ha-signer/src/db/postgres.ts:
```diff
@@ -72,24 +74,46 @@ export class PostgresSlashingProtectionDatabase implements SlashingProtectionDat
 
     if (result.rows.length === 0) {
-      // This shouldn't happen - the query always returns either the inserted or existing row
-      throw new Error('INSERT_OR_GET_DUTY returned no rows');
+      // this should never happen as the retry function should throw if it still fails after retries
+      throw new Error('INSERT_OR_GET_DUTY returned no rows after retries');
     }
```
I think we should panic if we get more than one row too (e.g. to catch bad DB constraints)

alexghr commented on yarn-project/validator-ha-signer/src/db/types.ts:
```diff
@@ -78,29 +83,27 @@ export interface ValidatorDutyRecord {
+  blockNumber: BlockNumber | CheckpointNumber;
+  /** The signing root (hash) for this duty */
   messageHash: string;
+  /** Unique identifier for the node that acquired the lock */
```
I'd remove the "unique" qualifier because the nodeID is user controlled so we can't prevent duplicates.

alexghr commented on yarn-project/validator-ha-signer/src/db/types.ts:
```diff
@@ -78,29 +83,27 @@ export interface ValidatorDutyRecord {
   validatorAddress: EthAddress;
-  slot: bigint;
+  slot: SlotNumber;
+  blockIndexWithinCheckpoint?: number;
```
Is there a risk nodes might request signatures for `blockIndexWithinCheckpoint: undefined` and `blockIndexWithinCheckpoint: 0` but both requests represent the same duty?

spypsy replied:
updated db methods to require `blockIndexWithinCheckpoint` to be defined and that's where we do the conversion
`blockIndexWithinCheckpoint ?? -1` so the two should always be distinguished

spypsy replied:
I think I misunderstood what you meant initially. `blockIndexWithinCheckpoint: undefined` should **not** be requested for `0`, that would be wrong. But can add strict type-checking on `BLOCK_PROPOSAL` which is the duty that strictly needs `blockIndexWithinCheckpoint`

alexghr commented on yarn-project/validator-ha-signer/src/config.ts:
```diff
@@ -20,65 +26,51 @@ export interface SlashingProtectionConfig {
+    env: 'VALIDATOR_HA_MAX_STUCK_DUTIES_AGE_MS',
     description: 'The maximum age of a stuck duty in ms',
     // hard-coding at current 2 slot duration. This should be set by the validator on init
     ...numberConfigHelper(72_000),
```
```suggestion
    ...numberConfigHelper(144_000),
```

alexghr commented on yarn-project/validator-ha-signer/src/migrations.ts:
```diff
@@ -30,12 +32,50 @@ export async function runMigrations(databaseUrl: string, options: RunMigrationsO
 
   const log = createLogger('validator-ha-signer:migrations');
 
+  const dbDir = join(__dirname, 'db');
```
I thought `__dirname` wasn't available in ESM files.

alexghr commented on yarn-project/validator-ha-signer/src/migrations.ts:
```diff
@@ -30,12 +32,50 @@ export async function runMigrations(databaseUrl: string, options: RunMigrationsO
     const appliedMigrations = await runner({
       databaseUrl,
-      dir: join(__dirname, 'db', 'migrations'),
+      dir: tempMigrationsDir,
```
why is a temp directory needed? Some operators might run the image on a read-only filesystem so this command wouldn't be able to run.

alexghr commented on yarn-project/validator-ha-signer/src/validator_ha_signer.test.ts:
```diff
@@ -330,16 +522,96 @@ describe('ValidatorHASigner', () => {
+          dutyType: sameDutyType,
+          blockIndexWithinCheckpoint: sameBlockIndex,
+          messageHash: MESSAGE_HASH.toString(),
+          nodeId: nodeIds[0], // Check with any nodeId, should return the same record
```
Could you explain what this means?
> Check with any nodeId, should return the same record

spypsy replied:
we're using `tryInsertOrGetExisting` for a record that should already have been created so it doesn't matter which node is trying to insert or get it. What matters is that `isNew` is false.
Will update the comment

alexghr commented on yarn-project/validator-ha-signer/src/validator_ha_signer.ts:
```diff
@@ -35,15 +35,15 @@ import type { SigningContext, SlashingProtectionDatabase } from './types.js';
     this.log = createLogger('validator-ha-signer');
 
-    if (!config.enabled) {
+    if (!config.haSigningEnabled) {
```
I think this should be controlled at a higher level (ie. in the validator or the `start_node.ts` script) not in the constructor.

spypsy replied:
it is basically controlled at the validator factory (`.new`) but makes sense to have it here as a guard too?

spypsy committed:
feat: ValidatorKeystore implementation for high-availability signer

Fixes [A-354](https://linear.app/aztec-labs/issue/A-354/integrate-slash-protection-in-signing-path)

Integrates the existing validator HA signer into validator-client via a new `HAKeyStore` wrapper, adds type-safe duty identifiers using discriminated unions, and includes integration testing with multiple validator nodes.

## What's New

### 1. HAKeyStore Implementation

**New File**: `yarn-project/validator-client/src/key_store/ha_key_store.ts`

Transparent wrapper around `ExtendedValidatorKeyStore` that adds HA coordination:

- Drop-in replacement for any `ExtendedValidatorKeyStore` implementation
- AUTH_REQUEST duties bypass HA coordination (safe to sign multiple times)
- Signs with all addresses concurrently using `Promise.allSettled()`, filtering out already-signed duties

```typescript
// With HA protection
const sigs = await haKeyStore.signTypedData(typedData, {
  slot: 100n,
  blockNumber: 50n,
  blockIndexWithinCheckpoint: 0,
  dutyType: DutyType.BLOCK_PROPOSAL,
});

// Without HA protection (AUTH_REQUEST)
const sigs = await haKeyStore.signTypedData(typedData, {
  dutyType: DutyType.AUTH_REQUEST,
});
```

### 2. Type-Safe Duty Identifiers

**Modified**: `yarn-project/validator-ha-signer/src/db/types.ts`

Replaced optional `blockIndexWithinCheckpoint?: number` with discriminated unions:

```typescript
type DutyIdentifier =
  | BlockProposalDutyIdentifier // MUST have blockIndexWithinCheckpoint >= 0
  | OtherDutyIdentifier; // Does NOT have this field

interface BlockProposalDutyIdentifier {
  validatorAddress: EthAddress;
  slot: SlotNumber;
  blockIndexWithinCheckpoint: number; // Required
  dutyType: DutyType.BLOCK_PROPOSAL;
}
```

Enforces correctness at compile-time and prevents invalid database primary keys.

### 3. Extended Duty Type Support

Added: `CHECKPOINT_PROPOSAL`, `GOVERNANCE_VOTE`, `SLASHING_VOTE`, `AUTH_REQUEST`

### 4. Integration Test

**New File**: `yarn-project/validator-client/src/validator.ha.integration.test.ts`

E2E test with 5 validators sharing a PostgreSQL database. Tests concurrent block proposals, attestations, checkpoint proposals, and slashing protection.

## Key Changes

### Cleanup Timeout Configuration

Made `maxStuckDutiesAgeMs` dynamically computed from actual slot duration:

```typescript
const haConfig = {
  ...config,
  maxStuckDutiesAgeMs:
    config.maxStuckDutiesAgeMs ??
    epochCache.getL1Constants().slotDuration * 2 * 1000,
};
```

### SigningContext Now Required

All `ValidatorKeyStore` signing methods now require `SigningContext` parameter. Updated `LocalKeyStore`, `NodeKeystoreAdapter`, `Web3SignerKeyStore`, and new `HAKeyStore`.

### Database Schema

Added `block_index_within_checkpoint` to primary key to support multiple block proposals per slot:

```sql
PRIMARY KEY (validator_address, slot, duty_type, block_index_within_checkpoint)
```

## Configuration

```bash
VALIDATOR_HA_DATABASE_URL=postgresql://user:pass@host:port/db
VALIDATOR_HA_SIGNING_ENABLED=true
VALIDATOR_HA_NODE_ID=validator-node-1
VALIDATOR_HA_POLLING_INTERVAL_MS=100        # Optional
VALIDATOR_HA_SIGNING_TIMEOUT_MS=3000        # Optional
VALIDATOR_HA_MAX_STUCK_DUTIES_AGE_MS=144000 # Optional (defaults to 2× slot duration)
```

## Notes

- HA is opt-in via `VALIDATOR_HA_SIGNING_ENABLED=true`
- Non-HA validators work unchanged
- Lock-free read operations
- Atomic lock acquisition using `INSERT ... ON CONFLICT` with retry logic (10ms, 20ms, 30ms backoff) to handle PostgreSQL transaction visibility edge cases in high-concurrency scenarios
